File: src/model/bin.ts
import { ColorObj } from "./color";
import { ContentObj } from "./content";
import { DTO } from "./dto";

export class BinObj extends DTO {
  private readonly _id: string;
  private _containerID: string;
  private _width: number;
  private _height: number;
  private _x: number;
  private _y: number;
  private _color: ColorObj;
  private _unit: string;
  private _content: Array<ContentObj>;

  public static Parse(d: any): BinObj {
    return new BinObj(d);
  }

  public static Empty(): BinObj {
    return new BinObj({});
  }

  public JSON(): object {
    return {
      id: this.id,
      container_id: this._containerID,
      width: this._width,
      height: this._height,
      column_start_x: this._x,
      column_start_y: this._y,
      color: this.color.JSON(),
      unit: this._unit,
      content: Array.from(
        this._content ? this._content : Array<ContentObj>(0),
        (d) => d.JSON()
      ),
    };
  }

  public Text(index: number): JSX.Element {
    return this._content[index].Text(this.unit);
  }

  public SearchText(index: number): string {
    return this._content[index].SearchText(this.unit);
  }

  private constructor(d: any) {
    super();

    this._id = d.id;
    this._containerID = d.container_id;
    this._width = d.width;
    this._height = d.height;
    this._x = d.column_start_x;
    this._y = d.column_start_y;
    this._color = ColorObj.Parse(d.color);
    this._unit = d.unit;
    this._content = Array.from(d.content ? d.content : Array<object>(0), (d) =>
      ContentObj.Parse(d)
    );
  }

  get id(): string {
    return this._id;
  }

  get containerID(): string {
    return this._containerID;
  }

  set containerID(value: string) {
    this._containerID = value;
  }

  get width(): number {
    return this._width;
  }

  set width(value: number) {
    if (value < 5) {
      throw new Error("width cannot be smaller then 5");
    }

    this._width = value;
  }

  get height(): number {
    return this._height;
  }

  set height(value: number) {
    if (value < 5) {
      throw new Error("height cannot be smaller then 5");
    }

    this._height = value;
  }

  get x(): number {
    return this._x;
  }

  set x(value: number) {
    if (value < 0) {
      throw new Error("x cannot be negative");
    }

    this._x = value;
  }

  get y(): number {
    return this._y;
  }

  set y(value: number) {
    if (value < 0) {
      throw new Error("x cannot be negative");
    }

    this._y = value;
  }

  get color(): ColorObj {
    return this._color;
  }

  set color(value: ColorObj) {
    this._color = value;
  }

  get unit(): string {
    return this._unit;
  }

  set unit(value: string) {
    this._unit = value;
  }

  get content(): Array<ContentObj> {
    return this._content;
  }

  set content(value: ContentObj[]) {
    this._content = value;
  }
}

File: app/model/bin.go
package model

import (
	"fmt"
	"time"

	"github.com/blevesearch/bleve/v2/mapping"
)

type Bin struct {
	//nolint:structcheck,unused
	tableName struct{} `pg:"goober.bin,alias:bin"`

	ID           string     `json:"id" pg:"id,pk,type:uuid"`
	ContainerID  string     `json:"container_id" pg:"container_id,type:uuid"`
	CreatedAt    time.Time  `json:"created_at" pg:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at" pg:"updated_at"`
	DeletedAt    *time.Time `json:"deleted_at" pg:"deleted_at,soft_delete"`
	Width        int        `json:"width" pg:"width,use_zero"`
	Height       int        `json:"height"  pg:"height,use_zero"`
	ColumnStartX int        `json:"column_start_x"  pg:"column_start_x,use_zero"`
	ColumnStartY int        `json:"column_start_y"  pg:"column_start_y,use_zero"`
	Color        Color      `json:"color"  pg:"color"`
	Unit         UnitType   `json:"unit"  pg:"unit,type:goober.unit_t"`
	Content      []Content  `json:"content" pg:"rel:has-many"`
}

type Content struct {
	//nolint:structcheck,unused
	tableName struct{} `pg:"goober.content,alias:content"`

	ID        string      `json:"id" pg:"id,pk,type:uuid"`
	BinID     string      `json:"bin_id" pg:"bin_id,fk,type:uuid"`
	CreatedAt time.Time   `json:"created_at" pg:"created_at"`
	UpdatedAt time.Time   `json:"updated_at" pg:"updated_at"`
	DeletedAt *time.Time  `json:"deleted_at" pg:"deleted_at,soft_delete"`
	Type      ContentType `json:"content_type"  pg:"content_type,type:goober.content_t"`
	Bolt      *Bolt       `json:"bolt,omitempty" pg:"rel:belongs-to"`
	Washer    *Washer     `json:"washer,omitempty" pg:"rel:belongs-to"`
	Screw     *Screw      `json:"screw,omitempty" pg:"rel:belongs-to"`
}

type IContent interface {
	fmt.Stringer
	// BleveType is the private BleveClassifier
	BleveType() string

	CreateMapping(fieldMapping *mapping.FieldMapping) (string, *mapping.DocumentMapping)
}
